core/dbt/parser/base.py

```diff
@@ -265,3 +265,43 @@ def parse_node(self, node, node_path, package_project_config, tags=None,
         parsed_node.validate()
 
         return parsed_node
+
+    def check_block_parsing(self, name, path, contents):
+        if not dbt.flags.WRAP_MODELS_IN_TAGS:
+            return
+
+        wrapped = ''.join(
+            ['{% model ', name, ' %}', contents, '{% endmodel %}']
+        )
+        blocks = dbt.clients.jinja.extract_toplevel_blocks(wrapped)
+        if len(blocks) == 0:
+            raise dbt.exceptions.InternalException(
+                'Failed to extract node {} ({}) from its wrapped block!'
+                .format(name, path)
+            )
+        elif len(blocks) > 1:
+            raise dbt.exceptions.InternalException(
+                'Found multiple blocks when parsing node {} ({}) from its '
+                'wrapped block, expected 1!'.format(name, path)
+            )
+        block = blocks[0]
+        if block.block_type_name != 'model':
+            raise dbt.exceptions.InternalException(
+                'Found invalid block type ("{}") when parsing node {} ({}) '
+                'from its wrapped block, expected "model"'
+                .format(block.block_type_name, name, path)
+            )
+
+        if block.block_name != name:
+            raise dbt.exceptions.InternalException(
+                'Found invalid block name ("{}") when parsing node {} ({}) '
+                'from its wrapped block, expected "{}"'
+                .format(block.block_name, name, path, name)
+            )
+
+        if blocks[0].contents != contents:
+            raise dbt.exceptions.InternalException(
+                'Found invalid block contents when parsing node {} ({}) '
+                'from its wrapped block, expected it to match original file'
+                .format(name, path)
+            )
```

Comment:
this is interesting, but it seems almost tautological.... unless someone included an unclosed comment or a `{% endmodel %}` in their model i would expect this to always pass. I am more interested in just running the toplevel block parser over the _unwrapped_ model contents, and comparing any compiler errors to those raised by the jinja parser. if the block parser raises an error, but the jinja parser does not, then we have a bug that we need to fix.

you should do the exact same thing for macro files too. `--wrap-models-in-tags` is probably a bad name if this turns the parser on for macros and models, so maybe `--test-new-parser` or something like that would be a better name

Reply:
>unless someone included an unclosed comment or a `{% endmodel %}` in their model i would expect this to always pass

Well, me too, but I thought the idea was to shake out bugs around top-level block extraction that _are_ in that category. I guess I misunderstood - I'll change it to be the way you've described, which seems like a reasonable enough test.

Reply:
@beckjake gotcha. I care more about comparing our parser to jinja's parser on arbitrary jinja code, not necessarily testing model block extraction. unrelated: i think wrapping model files in model blocks is a good way to achieve forwards compatibility with this new parser.

Reply:
Yup! I that is the longer-term plan, I think. Maybe we can get it in for wilt-chamberlain if there isn't too much already, that would be nice.